Major API refactor and additional features/fixes

andywaplinger opened the pull request:
Fixes #7, fixes #3.

**Major Changes**
- Updated API to MBTA API V3; V2 will be deprecated in April 2018
- Split API calls across multiple functions with deferred callbacks to get all required data
   - Requires the use of jQuery 3 for deferred callbacks
- Modified names of some subway stops to better reflect what they're typically called; users using affected stops will need to update the name in their config
- Added commuter rail stops and bus stops
- Added ability to switch been predictions and schedules using `predictedTimes` boolean
- Added ability to display departure time using `showDepartTime` boolean
   - Corrected the use of departure time vs arrival time
- Added `colorIcons` boolean so vehicle icons reflect their line color

**Minor Changes**
- Trips are sorted by ETA, not departure time
- Trips filtered by ETA when `maxTime` is set
- Added bus route number to trip description
- Fixed alerts and enabled marquee scrolling
- Added dictionary for trip directions (more info below)
- Added "mins" or "min" to ETA time so it's more human-readable
- Accounted for trips with no specified arrival or departure times
- Accounted for no alerts at a stop if alerts are turned on

**Known Issues**
- When there are multiple alerts on a trip, only the first alert is shown
- Direction filtering is not yet functional, but some groundwork has been laid

andywaplinger committed:
Major API refactor and additional features/fixes

edward-shen commented:
Thanks so much for the PR!

I'll do a proper review soon, but for now some quick notes:

Resolves #7. Does this resolve #4, #3, and #1? Did I forget anything?

I took a quick review at the changes with the API V3: looks like we can remove the `updateInterval` limit since each key can handle 1k request/minute. This is more of a note for myself, but you're free to make a change that limits it to 1 (split-second updating is a little unnecessary in all use cases).

I'm still quite new to Boston; what are some colloquial examples of subway stops? And can you tell me which ones were changed? (For documentation purposes)

Can you also update the readme for getting an API v3 key? I think we just need to change the old link to https://api-v3.mbta.com/register

andywaplinger commented:
You're welcome!

I'd say it definitely solves #3 and #7, and solves or, at least, helps #4. I'm doing more testing now and I wouldn't say that #1 is fixed. The alerts work better, but still only work well in full-width positions. I'll have to dig further into that.

I just tested setting the limit to 1s and quickly received 429 Too Many Requests. I'd recommend keeping `updateInterval`, but I can change it to 10s default.

Most of the names I changed had "Station" at the end when that's typically dropped, i.e. "Riverside Station" is now "Riverside", or "Boylston Street Station" is just "Boylston". Here's a sheet with the updated stations - [updated_stops.xlsx](https://github.com/edward-shen/MMM-MBTA/files/1657457/updated_stops.xlsx) 

I'll also update the URL in the readme.

edward-shen commented:
It's default limited to 10s, so you shouldn't need to worry about that.

edward-shen commented:
Made some minor changes, you might need to merge changes from master to resolve conflicts.

edward-shen commented on MMM-MBTA.js:
```diff
@@ -70,7 +73,28 @@ Module.register("MMM-MBTA", {
+        }
+        if (this.config.direction.includes("Eastbound")) {
+            this.filterDirection.push("1");
+        }
```
Wouldn't a switch case be better here? We can then use fall-through to make this a lot cleaner.

andywaplinger replied:
Sure, just like this? 
```js
switch (this.config.direction) {
    case "Southbound":
    case "Westbound":
    case "Outbound":
        this.filterDirection.push("0");
        break;
    case "Northbound":
    case "Eastbound":
    case "Inbound":
        this.filterDirection.push("1");
        break;
}
```

edward-shen replied:
Yup!

edward-shen commented on MMM-MBTA.js:
```diff
@@ -135,34 +153,98 @@ Module.register("MMM-MBTA", {
+                        break;
+                    default:
+                        descCell.innerHTML += " | " + routeId;
+                }
```
What are SL1-4 and CT1-3? I'm assuming CT# stands for Commuter Train, but what are SL1?

Can you add some comments?

andywaplinger replied:
Those are their public-facing route names. Aside from those, every other bus uses its `routeId` as it's public route name.

edward-shen commented on MMM-MBTA.js:
```diff
@@ -171,24 +253,51 @@ Module.register("MMM-MBTA", {
+                        arrTimeCell.innerHTML = "No arrival";
+                    } else {
+                        arrTimeCell.innerHTML = moment.unix(this.stationData[i].preArr).format("h:mm");
+                    }
```
I think doing this is cleaner:
```js
if (isNaN(this.stationData[i].preArr) === true) {
    arrTimeCell.innertHTML = "No arrival";
} else {
    if (config.timerFormat === 24) {
        arrTimeCell.innerHTML = moment.unix(this.stationData[i].preArr).format("H:mm");
    } else {
        arrTimeCell.innerHTML = moment.unix(this.stationData[i].preArr).format("h:mm");
    }
}

edward-shen replied:
What does no arrival mean? When does a mode of transportation show this?

andywaplinger replied:
Good point. Changed.

Sometimes `arrival_time` is `null`. If it's a bus/subway, I'm not sure why. But if it's a train, it's typically because it already arrived and is waiting to depart.

edward-shen replied:
Oh god, we need to check for `null`?

If that's the case, we can check nullity just by doing the following:
```js
if (!this.stationData[i].preArr) {
    ...
}
```
This is because `null` is a falsey value, so when you negative null, it returns true.

Thanks javascript.

andywaplinger replied:
Ahh, right, one of my favorite things about JS...

edward-shen commented on MMM-MBTA.js:
```diff
@@ -171,24 +253,51 @@ Module.register("MMM-MBTA", {
+                        depTimeCell.innerHTML = moment.unix(this.stationData[i].preDt).format("h:mm");
+                    }
+                }
+                row.appendChild(depTimeCell);
```
Same as above.

andywaplinger replied:
Sometimes `departure_time` also comes in `null`. No idea why.

edward-shen commented on MMM-MBTA.js:
```diff
@@ -211,48 +320,46 @@ Module.register("MMM-MBTA", {
+
+            function onlyUnique(value, index, self) { 
+                return self.indexOf(value) === index;
+            }
```
Unused function, please remove.

andywaplinger replied:
Slipped through

edward-shen commented on MMM-MBTA.js:
```diff
@@ -211,48 +320,46 @@ Module.register("MMM-MBTA", {
+                return self.indexOf(value) === index;
+            }
+
+            let uniqueAlerts = [...new Set(alertsArray)];
```
Actually, if you're only going to use `uniqueAlerts`, you can initalize `alertsArray` to be a new `Set` rather than an empty array. Then we can just push elements to the Set without worrying about duplication, eliminating the need to use a spread operator.

andywaplinger replied:
Makes sense. Updated.

edward-shen commented on MMM-MBTA.js:
```diff
@@ -211,48 +320,46 @@ Module.register("MMM-MBTA", {
-                descCell.className = "light small";
-                row.appendChild(descCell);
-                
+                row.appendChild(alertCell);
```
Can you add a config param to have the option to hide the alerts cell if it says "No alerts"? I feel like some would prefer not having the options to see no alerts as all, rather than text that says "No alerts".

edward-shen replied:
Wait if there's no alerts then it should never show "No alerts" because the length of `uniqueAlerts` would be empty. Does this do anything? (An example configuration would be appreciated).

andywaplinger replied:
You're right, it wouldn't show anything. I insert "No alerts" because I think from a UX perspective it makes sense to inform a user who has `showAlerts` turned on (and is expecting to see something) that there's nothing to see. Otherwise they might get the impression that alerts aren't working.

I could see adding a config to not show anything. That way it's an intentional user decision.

edward-shen replied:
Yeah, if you wanted to add this feature (e.g. check if uniqueAlerts is empty, then add in "No alerts" if a config is set to true), feel free to!

andywaplinger replied:
Just added!

edward-shen commented on MMM-MBTA.js:
```diff
@@ -311,67 +417,215 @@ Module.register("MMM-MBTA", {
+        url += "&include=stop,route,trip,alerts&sort=arrival_time";
+
+        if (!this.config.predictedTimes) {
+            url += "&page[limit]=25";
```
Could we set the page limit to `this.config.numEntries` + some constant?

What if the user wants (and god forbid) 100 entries?

andywaplinger replied:
You mean `maxEntries`?
I'll do that and slap an additional 10 on there.

edward-shen commented on MMM-MBTA.js:
```diff
@@ -311,67 +417,215 @@ Module.register("MMM-MBTA", {
+        if (!this.config.predictedTimes) {
+            url += "&page[limit]=25";
+            url += "&filter[min_time]=" + moment().format("HH:mm");
+            url += "&filter[max_time]=" + moment().add(3, 'h').format("HH:mm");
```
Can you please leave a comment explaining this is necessary, for documentation's sake?

andywaplinger replied:
Added 

> Page and time limits necessary because otherwise "schedules" endpoint gets every single schedule

edward-shen replied:
Sorry, I meant to clarify `moment.add(3, 'h')`, but that's also good to have too

andywaplinger replied:
Oh, gotcha. Does that comment cover enough ground for you? Or do you want me to another specifically about limiting the time to 3 hours from the current time?

edward-shen commented on MMM-MBTA.js:
```diff
@@ -311,67 +417,215 @@ Module.register("MMM-MBTA", {
+        alertUrl += "?api_key=" + this.config.apikey;
+
+        return alertUrl;
+    },
```
It might be a good idea to make some sort of enumeration type and merge `tripURL` and `routeURL` with this.

JS enumerations don't really exist, but there are ways to emulate them.

Regardless, I feel like making a switch statement would make reduce the amount of duplicated code.

andywaplinger replied:
Fixed using a single function.

edward-shen commented on MMM-MBTA.js:
```diff
@@ -311,67 +417,215 @@ Module.register("MMM-MBTA", {
+                preDt: "None",
+                preArr: "None",
+                preETA: "None"
+            });
```
I'm a little uncomfortable requiring our code to have a "default" state, because that implies that our code implicitly requires our `stationData` requires a minimum length of one--it's a code smell, and a very smelly one at that.

andywaplinger replied:
If there isn't one station, then `getDom()` skips the for loop that renders all trip data. I can include an if statement before the loop that catches when `stationData.length === 0` and renders just the `symbolCell` and `descCell`, unless you have a different idea.

andywaplinger replied:
Fixed - opted for the if statement.

edward-shen commented on MMM-MBTA.js:
```diff
@@ -311,67 +417,215 @@ Module.register("MMM-MBTA", {
         }
+        preDt = parseInt(moment(data.data[pred].attributes.departure_time).format("X"));
+        preArr = parseInt(moment(data.data[pred].attributes.arrival_time).format("X"));
+        preETA = moment(data.data[pred].attributes.arrival_time).diff(moment(), 'seconds') - 30; //Better safe than sorry?
```
Do we want to be truthful and inaccurate, or lie and "in good faith"?

andywaplinger replied:
Haha, why not both? There's already the disclaimer, which is good. We could also include that we've added a 30s buffer to help ensure they're on time.

andywaplinger committed:
Merged with master

andywaplinger committed:
Merged with master

andywaplinger commented:
Cool, I merged your changes. Should be good to go now.

Edit: made a commit error originally, but it should be fixed.

andywaplinger committed:
Merged with master

edward-shen commented on MMM-MBTA.js:
```diff
@@ -383,30 +637,36 @@ Module.register("MMM-MBTA", {
+                preDt: "None",
+                preArr: "None",
+                preETA: "None"
+            });
```
I would rather check if the stationData is empty when we're rendering, rather than making an "fake" station.

It's a misrepresentation of our data definition.

andywaplinger replied:
Fixed

edward-shen reviewed (commented):
If you could fix and/or explain some of the notes I've made, I would greatly appreciate it. Sorry it took so long, there was a lot more code than I expected and I got caught up with work.

edward-shen commented:
Can you make sure you're pushing to your changes to github? I can't see local git changes!

andywaplinger commented:
Oh yeah, sorry, I've been making a few of these changes before pushing. I'll push what I have now.

andywaplinger committed:
- Change `updateInterval` default to 10s
- Change `direction` filtering to switch case
- Clean up description, arrival, departure, and alerts
- Allow users to retrieve as many schedules as they want

andywaplinger committed:
Merge route, trip, and alert URL functions

andywaplinger committed:
- Removed needing default state
- Retrieve all alerts from all predictions
- Changed icon when no upcoming trips; icon is red when `colorIcons` is `true`

edward-shen commented on MMM-MBTA.js:
```diff
@@ -418,145 +438,110 @@ Module.register("MMM-MBTA", {
+        url += "?api_key=" + this.config.apikey;
 
-        return alertUrl;
+        return url
```
Missing semi-colon

andywaplinger replied:
Added

edward-shen commented on MMM-MBTA.js:
```diff
@@ -92,170 +106,283 @@ Module.register("MMM-MBTA", {
-                        preAwayCell.innerHTML = minutes + ":" + seconds;
-                    } else {
-                        preAwayCell.innerHTML = seconds;
+                    if (isNaN(this.stationData[i].routeId) === false) {
```
Same thing here, can just be
```js
if (this.stationData[i].routeId)  { ... }
```

andywaplinger replied:
What do you mean? Only bus routes use straight numbers. Everything else is alphanumeric.

I can replace it with `if ($.isNumeric(this.stationData[i].routeId))` instead so it's a little cleaner and clearer. Checking whether something is not _not_ a number is kinda dumb.

edward-shen commented on MMM-MBTA.js:
```diff
@@ -92,170 +106,283 @@ Module.register("MMM-MBTA", {
+                        var seconds = preETATime % 60;
+                        
+                        if (this.config.showMinutesOnly) {
+                            if (isNaN(minutes) === true) {
```
This also a null check?

andywaplinger replied:
Yeah. Just replaced it with `if (!minutes)`.

edward-shen commented on MMM-MBTA.js:
```diff
@@ -277,86 +404,199 @@ Module.register("MMM-MBTA", {
+                routeId = data.data[pred].relationships.route.data.id;
+                tripId = data.data[pred].relationships.trip.data.id;
+                alertIds = data.data[pred].relationships.alerts.data;
+debugger
```
extraneous debugger

andywaplinger replied:
Removed

edward-shen commented on MMM-MBTA.js:
```diff
@@ -277,86 +404,199 @@ Module.register("MMM-MBTA", {
+                alert = alertsParse[i].data.attributes.header;
+                alertsArray.push(alert);
+            }
+        };
```
extra semi colon

edward-shen replied:
Actually if you want this to look cleaner you can do a for...of loop:
```js
for (let alert of alertsParse) {
   if (alert.data) {
        alertsArray.push(alert.data.attributes.header);
    }
}
```

edward-shen replied:
Optional ofc, but nice learning point and could help make code cleaner in the future.

andywaplinger replied:
I'm all about cleaner code. I'll change it.

edward-shen reviewed (commented):
Just some last few style remarks!

One last thing: Can you provide me with a list of stations that show all your new features? It's for testing purposes, so I can triple check your code.

After I do that, then it should be on the home stretch!

edward-shen commented:
Would like to thank you again for such a large contribution to the repo; in fact, it looks like after this merge, you'll have more lines of code than me! :laughing: 

LMK when you're done with the last few changes, otherwise LGTM.

andywaplinger committed:
Add functionality to auto-hide Alerts section when empty

andywaplinger committed:
Minor code cleanup

andywaplinger commented:
You're very welcome! It was a really useful learning experience for me, and I think it's a great little app. I'm happy to help make it even better. 😄 

Let's merge this! 👍

andywaplinger committed:
Replaced double negative NaNs

andywaplinger commented:
For station testing, I've been using North Station (multiple subway lines and typically has one or two alerts), Central Square (typically has no alerts), South Station (buses, subways, and trains, sometimes with alerts), and random bus and commuter rail stops. Every feature should work, regardless of the stop. 😄

edward-shen commented:
Congratulations on your first PR! :tada:
